<center>
<img src="https://habrastorage.org/files/fd4/502/43d/fd450243dd604b81b9713213a247aa20.jpg">
    
## [mlcourse.ai](mlcourse.ai) – Open Machine Learning Course 

<center>Author: [Yury Kashnitskiy](http://yorko.github.io) <br>
Translated and edited by [Sergey Isaev](https://www.linkedin.com/in/isvforall/), [Artem Trunov](https://www.linkedin.com/in/datamove/), [Anastasia Manokhina](https://www.linkedin.com/in/anastasiamanokhina/), and [Yuanyuan Pao](https://www.linkedin.com/in/yuanyuanpao/) <br>All content is distributed under the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

# <center> Assignment #1 (demo)
## <center>  Exploratory data analysis with Pandas


**In this task you should use Pandas to answer a few questions about the [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) dataset. (You don't have to download the data – it's already here). Choose the answers in the [web-form](https://docs.google.com/forms/d/1uY7MpI2trKx6FLWZte0uVh3ULV4Cm_tDud0VDFGCOKg). This is a demo version of an assignment, so by submitting the form, you'll see a link to the solution .ipynb file.**

Unique values of all features (for more information, please see the links above):
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [1]:
import pandas as pd

In [3]:
DATA_URL = "https://raw.githubusercontent.com/Yorko/mlcourse.ai/main/data/"
data = pd.read_csv(DATA_URL + "adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. How many men and women (*sex* feature) are represented in this dataset?** 

In [4]:
data["sex"].value_counts()

sex
Male      21790
Female    10771
Name: count, dtype: int64

- Male: 21790
- Female:    10771

**2. What is the average age (*age* feature) of women?**

In [5]:
data[data["sex"] == "Female"]["age"].mean()

36.85823043357163

- average age: 36.85823043357163

**3. What is the percentage of German citizens (*native-country* feature)?**

In [6]:
(data.groupby("native-country").size() / len(data))["Germany"] * 100

0.42074874850281013

- German citizen percent: - German citizen percent: 

**4-5. What are the mean and standard deviation of age for those who earn more than 50K per year (*salary* feature) and those who earn less than 50K per year? **

In [7]:
mean_age = data[data["salary"] != "<=50K"]["age"].mean()
print(f"Средний возраст: {mean_age}")
std_age = data[data["salary"] != "<=50K"]["age"].std()
print(f"Средне квадратичное отклонение: {std_age}")

Средний возраст: 44.24984058155847
Средне квадратичное отклонение: 10.519027719851826


**6. Is it true that people who earn more than 50K have at least high school education? (*education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate* feature)**

In [16]:
data[(data["salary"] == ">50K") & (data["education"] == "10th")]

age  workclass  fnlwgt education  education-num      marital-status  \
588     52    Private   25826      10th              6  Married-civ-spouse   
1395    42    Private  230624      10th              6       Never-married   
1531    40    Private   85995      10th              6  Married-civ-spouse   
1533    49    Private  245305      10th              6  Married-civ-spouse   
3137    51    Private  252903      10th              6  Married-civ-spouse   
...    ...        ...     ...       ...            ...                 ...   
25527   58    Private  160662      10th              6  Married-civ-spouse   
25707   53  State-gov  153486      10th              6  Married-civ-spouse   
28111   48    Private  141511      10th              6  Married-civ-spouse   
30496   55    Private  134120      10th              6  Married-civ-spouse   
32179   58    Private  250206      10th              6  Married-civ-spouse   

              occupation relationship   race   sex  capital-gain  \
588         Craft-repair      Husband  White  Male             0   
1395    Transport-moving    Unmarried  White  Male             0   
1531   Machine-op-inspct      Husband  White  Male             0   
1533    Transport-moving      Husband  Black  Male             0   
3137               Sales      Husband  White  Male             0   
...                  ...          ...    ...   ...           ...   
25527       Craft-repair      Husband  White  Male             0   
25707   Transport-moving      Husband  White  Male             0   
28111    Exec-managerial      Husband  White  Male             0   
30496       Craft-repair      Husband  White  Male         99999   
32179       Craft-repair      Husband  White  Male          5178   

       capital-loss  hours-per-week native-country salary  
588            1887              47  United-States   >50K  
1395              0              40  United-States   >50K  
1531              0              40  United-States   >50K  
1533              0              42  United-States   >50K  
3137           1977              40  United-States   >50K  
...             ...             ...            ...    ...  
25527             0              40  United-States   >50K  
25707             0              40  United-States   >50K  
28111             0              45  United-States   >50K  
30496             0              40  United-States   >50K  
32179             0              40  United-States   >50K  

[62 rows x 15 columns]

Так как существуют люди с образованием 10 классов, получающие более 50 тысяч. Следует это противоречит гипотезе о том, что люди имеющие зп более 50 тысяч имеет хотя бы высшее образование.

**7. Display age statistics for each race (*race* feature) and each gender (*sex* feature). Use *groupby()* and *describe()*. Find the maximum age of men of *Amer-Indian-Eskimo* race.**

In [18]:
data.groupby("race").describe()

age                                                \
                      count       mean        std   min   25%   50%   75%   
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                           fnlwgt                 ... capital-loss          \
                     max    count           mean  ...          75%     max   
race                                              ...                        
Amer-Indian-Eskimo  82.0    311.0  120831.147910  ...          0.0  1980.0   
Asian-Pac-Islander  90.0   1039.0  159940.609240  ...          0.0  2457.0   
Black               90.0   3124.0  228013.124200  ...          0.0  4356.0   
Other               77.0    271.0  197124.191882  ...          0.0  2179.0   
White               90.0  27816.0  187298.064280  ...          0.0  4356.0   

                   hours-per-week                                         \
                            count       mean        std  min   25%   50%   
race                                                                       
Amer-Indian-Eskimo          311.0  40.048232  11.695364  3.0  40.0  40.0   
Asian-Pac-Islander         1039.0  40.127045  12.556816  1.0  40.0  40.0   
Black                      3124.0  38.422855  10.315646  1.0  37.0  40.0   
Other                       271.0  39.468635  11.143755  5.0  36.0  40.0   
White                     27816.0  40.689100  12.544796  1.0  40.0  40.0   

                                
                     75%   max  
race                            
Amer-Indian-Eskimo  40.0  84.0  
Asian-Pac-Islander  40.0  99.0  
Black               40.0  99.0  
Other               40.0  98.0  
White               45.0  99.0  

[5 rows x 48 columns]

In [20]:
data.groupby("sex").describe()

age                                                       fnlwgt  \
          count       mean        std   min   25%   50%   75%   max    count   
sex                                                                            
Female  10771.0  36.858230  14.013697  17.0  25.0  35.0  46.0  90.0  10771.0   
Male    21790.0  39.433547  13.370630  17.0  29.0  38.0  48.0  90.0  21790.0   

                       ... capital-loss         hours-per-week             \
                 mean  ...          75%     max          count       mean   
sex                    ...                                                  
Female  185746.311206  ...          0.0  4356.0        10771.0  36.410361   
Male    191771.449013  ...          0.0  3770.0        21790.0  42.428086   

                                                
              std  min   25%   50%   75%   max  
sex                                             
Female  11.811300  1.0  30.0  40.0  40.0  99.0  
Male    12.119755  1.0  40.0  40.0  49.0  99.0  

[2 rows x 48 columns]

Максимальный возраст человека с расой Amer-Indian-Eskimo составляет 82 года.

**8. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (*marital-status* feature)? Consider as married those who have a *marital-status* starting with *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.**

In [21]:
data[(data["sex"] == "Male") & (data["marital-status"].str.startswith("Married"))][
    "salary"
].value_counts(normalize=True)

salary
<=50K    0.559486
>50K     0.440514
Name: proportion, dtype: float64

In [22]:
data[(data["sex"] == "Male") & ~(data["marital-status"].str.startswith("Married"))][
    "salary"
].value_counts(normalize=True)

salary
<=50K    0.915505
>50K     0.084495
Name: proportion, dtype: float64

Среди мужчин которые не женаты процент, зарабатывающих больше 50 тысяч, больше.

**9. What is the maximum number of hours a person works per week (*hours-per-week* feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?**

In [24]:
max_time = data["hours-per-week"].max()
print(f"Максимальное время работы: {max_time}")
count_people_with_max_time = len(data[data["hours-per-week"] == max_time])
print(f"Количество людей с максимальной занятостью: {count_people_with_max_time}")
percente_of_rich = (
    len(data[(data["hours-per-week"] == max_time) & (data["salary"] == ">50K")])
    * 100
    / count_people_with_max_time
)
print(f"Процент богатых людей с максимальной занятостью : {percente_of_rich}")

Максимальное время работы: 99
Количество людей с максимальной занятостью: 85
Процент богатых людей с максимальной занятостью : 29.41176470588235


**10. Count the average time of work (*hours-per-week*) for those who earn a little and a lot (*salary*) for each country (*native-country*). What will these be for Japan?**

In [27]:
data.groupby(["native-country", "salary"]).describe()

age                                                  \
                        count       mean        std   min    25%   50%    75%   
native-country salary                                                           
?              <=50K    437.0  36.853547  12.532401  17.0  28.00  35.0  43.00   
               >50K     146.0  44.328767  10.158814  22.0  37.00  43.0  51.00   
Cambodia       <=50K     12.0  35.666667  12.404789  18.0  27.75  34.5  37.75   
               >50K       7.0  41.428571   7.828519  27.0  38.50  42.0  47.00   
Canada         <=50K     82.0  41.012195  16.507009  17.0  27.00  38.0  53.00   
...                       ...        ...        ...   ...    ...   ...    ...   
United-States  >50K    7171.0  44.297169  10.485946  19.0  37.00  44.0  51.00   
Vietnam        <=50K     62.0  33.951613  13.724965  19.0  23.00  30.0  43.75   
               >50K       5.0  35.400000  11.326959  23.0  30.00  30.0  43.00   
Yugoslavia     <=50K     10.0  38.000000  13.072448  20.0  26.50  41.0  44.50   
               >50K       6.0  40.166667  13.136463  29.0  35.00  35.5  39.00   

                             fnlwgt                 ... capital-loss          \
                        max   count           mean  ...          75%     max   
native-country salary                               ...                        
?              <=50K   90.0   437.0  195354.379863  ...          0.0  2339.0   
               >50K    76.0   146.0  187154.513699  ...          0.0  3004.0   
Cambodia       <=50K   65.0    12.0  167065.666667  ...          0.0  1876.0   
               >50K    50.0     7.0  237677.000000  ...          0.0     0.0   
Canada         <=50K   80.0    82.0  169279.585366  ...          0.0  2339.0   
...                     ...     ...            ...  ...          ...     ...   
United-States  >50K    90.0  7171.0  187433.992330  ...          0.0  3683.0   
Vietnam        <=50K   73.0    62.0  165028.725806  ...          0.0  2457.0   
               >50K    51.0     5.0  141648.400000  ...          0.0     0.0   
Yugoslavia     <=50K   56.0    10.0  227306.800000  ...          0.0     0.0   
               >50K    66.0     6.0  197059.166667  ...          0.0     0.0   

                      hours-per-week                                           \
                               count       mean        std   min    25%   50%   
native-country salary                                                           
?              <=50K           437.0  40.164760  12.817002   1.0  40.00  40.0   
               >50K            146.0  45.547945  10.202827   8.0  40.00  40.0   
Cambodia       <=50K            12.0  41.416667   3.088346  40.0  40.00  40.0   
               >50K              7.0  40.000000   0.000000  40.0  40.00  40.0   
Canada         <=50K            82.0  37.914634  13.012056   8.0  32.75  40.0   
...                              ...        ...        ...   ...    ...   ...   
United-States  >50K           7171.0  45.505369  11.025092   1.0  40.00  40.0   
Vietnam        <=50K            62.0  37.193548  12.422664   1.0  40.00  40.0   
               >50K              5.0  39.200000   1.788854  36.0  40.00  40.0   
Yugoslavia     <=50K            10.0  41.600000  11.305849  20.0  40.00  40.0   
               >50K              6.0  49.500000  11.202678  40.0  40.00  46.0   

                                   
                        75%   max  
native-country salary              
?              <=50K   45.0  99.0  
               >50K    50.0  80.0  
Cambodia       <=50K   40.5  50.0  
               >50K    40.0  40.0  
Canada         <=50K   40.0  70.0  
...                     ...   ...  
United-States  >50K    50.0  99.0  
Vietnam        <=50K   40.0  77.0  
               >50K    40.0  40.0  
Yugoslavia     <=50K   40.0  66.0  
               >50K    58.0  65.0  

[82 rows x 48 columns]

In [33]:
data[data["native-country"]=="Japan"].groupby(["native-country", "salary"]).describe()

age                                                 \
                      count       mean       std   min    25%   50%    75%   
native-country salary                                                        
Japan          <=50K   38.0  35.315789  9.392652  19.0  29.00  34.0  41.00   
               >50K    24.0  42.875000  6.771504  33.0  37.75  41.0  47.25   

                            fnlwgt                 ... capital-loss          \
                        max  count           mean  ...          75%     max   
native-country salary                              ...                        
Japan          <=50K   61.0   38.0  197242.526316  ...          0.0  1590.0   
               >50K    56.0   24.0  167433.708333  ...          0.0  1977.0   

                      hours-per-week                                          \
                               count       mean        std   min   25%   50%   
native-country salary                                                          
Japan          <=50K            38.0  41.000000  11.902759  10.0  40.0  40.0   
               >50K             24.0  47.958333  16.120414  21.0  40.0  42.5   

                                   
                        75%   max  
native-country salary              
Japan          <=50K   41.5  65.0  
               >50K    50.0  99.0  

[2 rows x 48 columns]

Japan >50:47.958333
Japan <=50:41